{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "aquatic-dutch",
   "metadata": {},
   "source": [
    "# Machine Learning For Genomics\n",
    "\n",
    "### Introduction\n",
    "Genomic data generated by researchers has grown exponentially. This demands even better tools to derive insights from the data, including augmenting with other data for better inference and decision-making. Machine learning, Deep learning, and artificial intelligence have matured with powerful tools that can be applied in Genomics. However, in Africa, there is still a skills gap among Bioinformatics students in these technologies. Although the MSc Bioinformatics curricula contain the course, it is not being taught, putting the students at a disadvantage, as Bioinfromatuics leans more towards data science. In this course, we introduce the basics of machine learning, including practical skills in transforming genomic data for machine learning modeling. \n",
    "\n",
    "### Competencies\n",
    "In this short course, we intend to impart knowledge and skills in the following competencies ([See ISCB Competencies](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005772)):\n",
    "1. Knowledge and skills: Details of the scientific discovery process and the role of bioinformatics in it.\n",
    "2. Knowledge, comprehension, and Application: Statistical, machine learning, and data science research methods in the context of molecular biology, genomics, medical or population genetics research\n",
    "3. Knowledge and Application: Command line and scripting-based computing skills appropriate to the discipline.\n",
    "4. Knowledge and skills: Data management\n",
    "\n",
    "### Learning Objectives\n",
    "To attain the above competencies, the workshop participants should be able to:\n",
    "1. Describe the application of machine learning in genomics\n",
    "2. Explain the various machine learning principles and how they can be applied to genomics\n",
    "3. Explain the research design approaches as applied to machine learning for genomics\n",
    "4. Know the various open science tools (Jupyter Notebooks, Pandas, Conda)and how they support a reproducible bioinformatics research\n",
    "5. Know the various machine learning frameworks in Python\n",
    "\n",
    "\n",
    "### Learning Outcomes\n",
    "From the above objectives, the workshop participant should acquire the following skills;\n",
    "1. Be able to set up Jupyter and Conda environments for a machine learning for a genomic project to ensure reproducibility\n",
    "2. Be able to transform genomic data for machine learning modeling\n",
    "3. Be able to perform exploratory analysis on genomic data, feature engineering, and parameter selection\n",
    "4. Be able to develop and validate machine learning models using genomic data\n",
    "\n",
    "### Instructors\n",
    "1. Caleb Kibet\n",
    "2. \n",
    "\n",
    "### Who should attend?\n",
    "\n",
    "EANBiT Fellows\n",
    "\n",
    "### Training Topics\n",
    "- \n",
    "\n",
    "# Quick Introduction to Jupyter Notebooks\n",
    "\n",
    "Throughout this course, we will be using Jupyter Notebooks. Although the HPC you will be using will have Jupyter setup, these notes are provided for you want to set it up in your Computer. \n",
    "\n",
    "## Introduction\n",
    "The Jupyter Notebook is an interactive computing environment that enables users to author notebooks, which contain a complete and self-contained record of a computation. These notebooks can be shared more efficiently. The notebooks may contain:\n",
    "* Live code\n",
    "* Interactive widgets\n",
    "* Plots\n",
    "* Narrative text\n",
    "* Equations\n",
    "* Images\n",
    "* Video\n",
    "\n",
    "It is good to note that \"Jupyter\" is a loose acronym meaning Julia, Python, and R; the primary languages supported by Jupyter. \n",
    "\n",
    "The notebook can allow a computational researcher to create reproducible documentation of their research. As Bioinformatics is datacentric, use of Jupyter Notebooks increases research transparency, hence promoting open science. \n",
    "\n",
    "## Pre-requisites\n",
    "The machine learning for genomics assumes farmiliarity with Python and Pandas. Please have a look at the Python4Bioinformatics training materials for a refresher. \n",
    "\n",
    "## First Steps\n",
    "\n",
    "### Installation\n",
    "\n",
    "1. [Download Miniconda](https://www.anaconda.com/download/) for your specific OS to your home directory\n",
    "    - Linux: `wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh`\n",
    "    - Mac: `curl https://repo.continuum.io/miniconda/Miniconda3-latest-MacOSX-x86_64.sh`\n",
    "2. Run:\n",
    "    - `bash Miniconda3-latest-Linux-x86_64.sh`\n",
    "    - `bash Miniconda3-latest-MacOSX-x86_64.sh`\n",
    "3. Follow all the prompts: if unsure, accept defaults\n",
    "4. Close and re-open your terminal\n",
    "5. If the installation is successful, you should see a list of installed packages with\n",
    "    - `conda list`\n",
    "\n",
    "If the command cannot be found, you can add Anaconda bin to the path using:\n",
    "    `export PATH=~/miniconda3/bin:$PATH`\n",
    "\n",
    "For reproducible analysis, you can [create a conda environment](https://conda.io/docs/user-guide/tasks/manage-environments.html) with all the Python packages you used.\n",
    "\n",
    "    `conda create --name bioinf python jupyter`\n",
    "    \n",
    "To activate the conda environment:\n",
    "\n",
    "    `source activate bioinf`\n",
    "\n",
    "Having set-up conda environment, you can install `jupyter lab` using pip. \n",
    "\n",
    "`conda install -c conda-forge jupyterlab`\n",
    "\n",
    "or by using pip\n",
    "\n",
    "`pip3 install jupyter`\n",
    "\n",
    "## How to learn from this resource?\n",
    "\n",
    "Download all the notebooks from [MachineLearning4Genomics(https://github.com/mbbu/MachineLearning4Genomics). The easiest way to do that is to clone the GitHub repository to your working directory using any of the following commands:\n",
    "\n",
    "    git clone https://github.com/mbbu/MachineLearning4Genomics.git\n",
    "\n",
    "or\n",
    "\n",
    "    wget https://github.com/mbbu/MachineLearning4Genomics/archive/master.zip\n",
    "    \n",
    "    unzip master.zip\n",
    "    \n",
    "    rm master.zip\n",
    "    \n",
    "    cd MachineLearning4Genomics-master\n",
    "    \n",
    "Then you can quickly launch jupyter lab using:\n",
    "\n",
    "`jupyter lab`\n",
    "\n",
    "NB: We will use a jupyter lab for training. \n",
    "A Jupyter notebook is made up of many cells. Each cell can contain Python code. You can execute a cell by clicking on it and pressing `Shift-Enter` or `Ctrl-Enter` (run without moving to the next line). \n",
    "\n",
    "\n",
    "## Contents\n",
    "\n",
    "This course is broken up into a number of notebooks (lectures).\n",
    "\n",
    "### Session 1\n",
    "* [00](Intro-to-Python/00.ipynb) This introduction with additional information below on how to get started in running python\n",
    "* [01](Intro-to-Python/01.ipynb) Basic data types and operations (numbers, strings) \n",
    "\n",
    "### Session 2\n",
    "* [02](Intro-to-Python/02.ipynb) String manipulation \n",
    "* [03](Intro-to-Python/03.ipynb) Data structures: Lists and Tuples\n",
    "* [04](Intro-to-Python/04.ipynb) Data structures (continued): dictionaries\n",
    "\n",
    "### Session 3\n",
    "* [05](Intro-to-Python/05.ipynb) Control statements: if, for, while, try statements\n",
    "* [06](Intro-to-Python/06.ipynb) Functions\n",
    "* [07](Intro-to-Python/07.ipynb) Scripting with python\n",
    "* [08](Intro-to-Python/08.ipynb) Data Analysis and plotting with Pandas\n",
    "* [09](Intro-to-Python/09.ipynb) Reproducible Bioinformatics Research\n",
    "\n",
    "This is a tutorial style introduction to Python. For a quick reminder/summary of Python syntax, the following [Quick Reference Card](http://www.cs.put.poznan.pl/csobaniec/software/python/py-qrc.html) may be useful. A longer and more detailed tutorial style introduction to python is available from the python site at: https://docs.python.org/3/tutorial/.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### Resources to use:\n",
    "\n",
    "1. [Encoding DNA ](https://medium.com/mlearning-ai/apply-machine-learning-algorithms-for-genomics-data-classification-132972933723)\n",
    "2. [Machine Learning in Bioinformatics: Genome Geography:](https://towardsdatascience.com/machine-learning-in-bioinformatics-genome-geography-d1b1dbbfb4c2)From raw sequencing reads to a machine learning model, which infers an individuals geographical origin based on their genomic variation.\n",
    "3. [Deep Learning for Genomics](https://colab.research.google.com/github/TankMermaid/1000-genomes-genetic-maps/blob/master/A_Primer_on_Deep_Learning_in_Genomics_Public.ipynb#scrollTo=bzsbNHqWiFek)\n",
    "4. [Machine Learning for Genomics](https://towardsdatascience.com/machine-learning-for-genomics-c02270a51795). How to transform your genomics data to fit into machine learning models. \n",
    "5. [Machine Learning For Good](https://github.com/DeltaAnalytics/machine_learning_for_good)\n",
    "\n",
    "\n",
    "## How to Contribute\n",
    "\n",
    "To contribute, fork the repository, make some updates and send me a pull request. \n",
    "\n",
    "Alternatively, you can open an issue. \n",
    "\n",
    "## License\n",
    "This work is licensed under the Creative Commons Attribution 3.0 Unported License. To view a copy of this license, visit http://creativecommons.org/licenses/by/3.0/"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
